features/101-design-webhooks/design.md

```diff
@@ -0,0 +1,129 @@
+# next webhooks design
+
+## Overview
+
+skygear sends webhooks for events that happen in your app, and are designed to help a developer to monitor events, sync database, and many other use cases.
+
+When one of events is triggered, skygear sends a HTTP POST with a payload to the webhook's configured URL.
+
+## SYNC and ASYNC hooks
+
+There are two types of webhook:
+
+- `SYNC` - gear's operation will be blocked until a response is received. Gear should respect `SYNC` hook's response, if it indicates failed, operation should abort and return a formatted error response to the client.
+- `ASYNC` - gear's operation won't be blocked and the response from the webhook is ignored.
+
+## Request Headers
+
+Following headers will be added to webhook POST request:
+
+- `X-Skygear-Webhook-Req-ID`: an ID of the request.
+- `X-Skygear-Webhook-Async`: if is a `async` hook, then the value is `TRUE` otherwise is `FALSE`. 
+- `X-Skygear-Webhhok-Signature`: a signature string of the webhook event, a develop should can verify the request by skygear's SDK or manually.
+
+## Request Payload
+
+Each event type has a payload which carries relevant information, and each payload data is defined by each gear.
+
+All event payloads follows following payload format:
+
+```json=
+{
+    "event": "event_name",
+    "data": {
+        "key_name_1": "event_data_key_value",
+        "key_name_2": "event_data_key_value",
+        ...,
+        "context": {
+            "user": <current user object>,
+            "req": {
+                "path": "request_path",
+                "body": <original_request_body>,
+                "id": "request_id"
+            }
+        }
+    }
+}
+```
+
+Note that, `context` is above format is still under discussion.
```

Comment:
should it be:

{
  "event": xxx,
  "data": {
    "context": { } 
     ... 
  }
}

or

{
  "event": ...,
  "data": { ... },
  "context": { ... }
}

Reply:
In https://github.com/SkygearIO/features/pull/261, we have hooked cloud function signature as: `before_password_changed_sync(user, original_user, context)`. 

Here I would like to keep the same signature for web hook, so the request payload is defined like:
```
{
    "event": "before_password_changed",
    "data": {
        "user": <user object>,
        "original_user": <user object>,
        "context": <context object>
    }
}
```
`data` attribute is one-one mapping to hook arguments. A Developer can easily understand the pattern and use web hook like using hooked cloud function.